### PR TITLE
George Ho

### PR DESCRIPTION
**Pull request by georgehwho**

### CODE REVIEW

File: lib/product.rb [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+    @unit_price = unit_price
+    @quantity = quantity.to_i
+  end
+
+  # returns the sum of the unit price multiplied by the quantity
+  # no argument -> float
+  def total_price
+    unit_price * quantity
+  end
+
```
Comment (memcmahon) [inline lib/product.rb:16]: Oh no!  you're missing two methods from iteration 1 😢

File: lib/shopping_cart.rb [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+  # tells you what the cart's name is and what the cart's total capacity is
+  # no argument -> hash
+  def details
+    {name: name, capacity: capacity }
+  end
+
+  # tells you the total # of products within a cart
+  # no argument -> integer
+  def total_number_of_products
+    products.reduce(0) { |memo, product| memo += product.quantity }
```
Comment (memcmahon) [inline lib/shopping_cart.rb:25]: You could also use `sum` here

File: lib/shopping_cart.rb [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+  # tells you whether the cart is full or not
+  # no argument -> boolean
+  def is_full?
+    total_number_of_products < capacity ? false : true
+  end
+
+  # returns an array of all products of a category
+  # category -> array of products in that category
+  def products_by_category(category)
+    product_breakdown[category]
```
Comment (memcmahon) [inline lib/shopping_cart.rb:37]: great re-use of `product_breakdown`!

File: lib/shopping_cart.rb [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+  # returns the % of the cart that is occupied
+  # no argument -> float of % occupied
+  def percentage_occupied
+    ratio = total_number_of_products / capacity.to_f * 100
+    ratio.round(2)
+  end
+
+  # sorts the products by their quantity
+  # no argument -> array of products sorted by quantity
+  def sorted_product_by_quantity
```
Comment (memcmahon) [inline lib/shopping_cart.rb:49]: keep a close eye on the interaction pattern - this should be `sorted_products_by_quantity`, and we wanted it in decrementing order (or, the reverse of the standard sort_by)

File: lib/shopping_cart.rb [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+    products.sort_by(&:quantity)
+  end
+
+  # returns a hash of all the producted sorted by their category in alphabetical order
+  # no argument -> hash of alphabetically & catagorically sorted products
+  def product_breakdown
+    products.sort_by(&:category).group_by {|product| product.category }
```
Comment (memcmahon) [inline lib/shopping_cart.rb:56]: I don't think we need the products sorted 🧐

**Review by memcmahon — COMMENTED** [review]
Good job - take a look at my notes and let me know if you have any questions!

Score: 3.5

File: lib/product.rb (hunk 1 above)
Comment (georgehwho) [reply lib/product.rb:16]: uh oh lol that's totally on me I completely missed those two methods 🥲

File: lib/shopping_cart.rb (hunk 2 above)
Comment (georgehwho) [reply lib/shopping_cart.rb:25]: oh good to know! that would've saved me a ton of typing.

File: lib/shopping_cart.rb (hunk 3 above)
Comment (georgehwho) [reply lib/shopping_cart.rb:37]: thanks!

File: lib/shopping_cart.rb (hunk 4 above)
Comment (georgehwho) [reply lib/shopping_cart.rb:49]: ahhh good catch! I need to pay way more attention to detail next time.

File: lib/shopping_cart.rb (hunk 5 above)
Comment (georgehwho) [reply lib/shopping_cart.rb:56]: ah I was basing it off of the output from the page 
```
pry(main)> cart.product_breakdown
#=> {:meat=>[#<Product:0x00007fccd2985f48...>], :paper=> [#<Product:0x00007fccd29b5720...>, #<Product:0x00007fccd2985f51...>], :produce=> [#<Product:0x00007fccd2985f53...>]}
```
which seemed to have the keys sorted alphabetically.

**Pull request closed without merging**